### ISSUE TITLE
Extend `fa::attention_forward` to support a **padding mask** that disables attention to masked key positions before softmax.

### ISSUE THREAD

**Issue by sreehariam** [sreehariam/fa-research-cpp#3]
The existing function signature remains:
```cpp
Tensor attention_forward(const Tensor& Q,
                         const Tensor& K,
                         const Tensor& V,
                         const Tensor* mask,
                         const AttentionOpts& opts);`

**Mask Semantics**

- Shape: (B, 1, 1, N), broadcast over all heads and query indices.
- Values: non-zero → keep, zero → masked.
- Masked logits must be replaced by −∞ before softmax.
- If all keys in a row are masked, the output for that row must be exact zeros.

**Acceptance Criteria**

- Supports mask shape (B, 1, 1, N); throws std::invalid_argument otherwise.
- Output remains finite; identical to unmasked output when mask = 1s.
- Deterministic small test → correct mean of unmasked V rows.
- All-masked rows return exact zeros.
- Causal attention still disabled (opts.causal=false).

**Out of Scope**

- Causal masking and CUDA support.
